compact.go

```diff
@@ -520,20 +536,33 @@ func (c *LeveledCompactor) write(dest string, meta *BlockMeta, blocks ...BlockRe
 	if err != nil {
 		return errors.Wrap(err, "open index writer")
 	}
-	defer indexw.Close()
+	writers = append(writers, indexw)
 
 	if err := c.populateBlock(blocks, meta, indexw, chunkw); err != nil {
 		return errors.Wrap(err, "write compaction")
 	}
+
+	// Compaction was canceled so remove tmp folders and return early.
+	select {
+	case <-c.ctx.Done():
+		for _, w := range writers {
+			w.Close()
+		}
+		if err := os.RemoveAll(tmp); err != nil {
+			level.Error(c.logger).Log("msg", "removed tmp folder after canceled compaction", "err", err.Error())
+		}
+		return ErrCompactionCanceled
+	default:
+	}
+
 	// We are explicitly closing them here to check for error even
 	// though these are covered under defer. This is because in Windows,
 	// you cannot delete these unless they are closed and the defer is to
 	// make sure they are closed if the function exits due to an error above.
-	if err = chunkw.Close(); err != nil {
-		return errors.Wrap(err, "close chunk writer")
-	}
-	if err = indexw.Close(); err != nil {
-		return errors.Wrap(err, "close index writer")
+	for _, w := range writers {
+		if err := w.Close(); err != nil {
+			return err
```

Comment:
We need `multierr` here ideally - same for the closing in `defer`

Reply:
And then we can always empty `writers` slice to not close twice

Reply:
very good suggestion , thanks!